### ISSUE TITLE
Subclass __init__ not type checked to pass parent classmethods

### ISSUE THREAD

**Issue by leonardgerardatomicmachinescom** [microsoft/pyright#9870]
**Describe the bug**

When a classmethod create an object of the class using `cls(arguments)`, the arguments are checked against the __init__ function of the class where the classmethod is defined and will correctly catch type errors. But if a class inherit from this class, it can define a non compatible __init__ function and we end up with runtime type error.

**Code**

```python
class Device():
    def __init__(self, i: int):
        print(i+1)

    @classmethod
    def from_config(cls, config: int) -> Self:
        """Create a device from a configuration."""
        return cls(config)

class DL(Device):
    def __init__(self, ids: list[int]):
        print(ids.append(1))

d = DL.from_config(4)  # runtime exception AttributeError: 'int' object has no attribute 'append'
d = DL.from_config([4])  # pyright typing exception (as expected)
```

I would expect pyright, at least in strict mode, to complain about the __init__ defined in DL.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

**Comment by erictraut** [comment]
Pyright doesn't enforce the LSP (Liskov Substitution Principle) for constructor methods. This is an intentional choice made by pyright, mypy, and other static type checkers for Python. There has been some discussions in the typing forum of changing this, but the consensus in the community is that this would result in too many false positive errors in typical Python code. You're correct that this does allow for a false negative in some cases, but that's a tradeoff that we need to make to avoid too many false positives.

In any case, this isn't a bug but rather an intentional design choice, so I'm going to close it as "won't fix".

If you'd like to argue in favor of changing the typing spec to mandate LSP enforcement for constructors, the [Python typing forum](https://discuss.python.org/c/typing/32) is the best place to do that. Just be warned that you're likely to get a lot of pushback from the community here.

Something that I might consider is adding a separate check for this condition (e.g. `reportIncompatibleConstructorOverride`), but I don't think that it could be enabled by default even in strict mode. It would simply be too noisy. I'm generally reluctant to add new checks that cannot be enabled by default in strict mode because they are difficult to discover and therefore get almost no usage.

**Comment by leonardgerardatomicmachinescom** [comment]
I am not looking for a general check of LSP for constructors, aka `reportIncompatibleConstructorOverride`. I would not enable such a check.

I am looking for something telling me that `DL.from_config` will call `DL.__init__` with wrong parameter types.

I would expect pyright to keep track of classmethods and be able to check they are valid in any class they are callable from (so one could override a class method to deal with a change in `__init__` signature).


Note that it is kind of unintuitive that pyright would not catch the error from my example, when it does catch it in this case:
```
class Device():
    def __init__(self, i: int):
        print(i+1)

    @classmethod
    def from_config(cls, config: list[int]) -> Self:
        """Create a device from a configuration."""
        return cls(config)
```

Pyright will error in this case with
> Argument of type "list[int]" cannot be assigned to parameter "i" of type "int" in function "__init__"
  "list[int]" is not assignable to "int"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)

**Comment by erictraut** [comment]
What you're asking for would require global program analysis or other complex and expensive analysis techniques. That's not something pyright (or AFAIK any other Python static type checker) does.

Pyright performs local analysis based on interface contracts defined at function boundaries. The Python type system is designed to accommodate local analysis. In your example, the method `from_config` is analyzed locally based on its interface contract and the context of its class. It is not re-analyzed from the context of derived classes.

**Comment by leonardgerardatomicmachinescom** [comment]
The interface contract defined at the classmethod boundaries, returning Self, is not correct if it assumes LSP for the constructors, but it is then not enforced.

So maybe we could have a check that report constructor LSP violation only when there is a class method. I understand this would mean storing an extra bit of information and that is not part of python class type interface today. But isn't pyright already storing more information than the written signature? (like when the return type of a function is omitted).
